src/sql/SqlService.ts

```diff
@@ -0,0 +1,422 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {SqlResult, SqlResultImpl} from './SqlResult';
+import {ConnectionManager, ConnectionRegistry} from '../network/ConnectionManager';
+import {SqlExpectedResultType, SqlStatement, SqlStatementOptions} from './SqlStatement';
+import {HazelcastSqlException, IllegalArgumentError} from '../core';
+import {SqlErrorCode} from './SqlErrorCode';
+import {SqlQueryId} from './SqlQueryId';
+import {SerializationService} from '../serialization/SerializationService';
+import {SqlExecuteCodec} from '../codec/SqlExecuteCodec';
+import * as Long from 'long';
+import {InvocationService} from '../invocation/InvocationService';
+import {ClientMessage} from '../protocol/ClientMessage';
+import {Connection} from '../network/Connection';
+import {SqlRowMetadataImpl} from './SqlRowMetadata';
+import {SqlCloseCodec} from '../codec/SqlCloseCodec';
+import {SqlFetchCodec, SqlFetchResponseParams} from '../codec/SqlFetchCodec';
+import {
+    tryGetArray,
+    tryGetBoolean,
+    tryGetEnum,
+    tryGetLong,
+    tryGetNumber,
+    tryGetString
+} from '../util/Util';
+import {SqlPage} from './SqlPage';
+import {HzLocalTime, HzLocalDate, HzLocalDateTime, HzOffsetDateTime} from './DatetimeClasses';
+
+/**
+ * SQL Service of the client. You can use this service to execute SQL queries.
+ *
+ * The service is in beta state. Behavior and API might change in future releases.
+ *
+ * ### Overview
+ * Hazelcast is able to execute distributed SQL queries over the following entities:
+ * * IMap
+ *
+ * During optimization, a statement is converted into a directed acyclic graph (DAG) that is sent to cluster members
```

Comment:
We can remove the this paragraph too, since it is server-side implementation detail

Reply:
what about 

Every table backed by an IMap has a set of columns that are resolved automatically. Column resolution uses IMap entries
 located on the member that initiates the query. The engine extracts columns from a key and a value and then merges them into a single column set. In case the key and the value have columns with the same name, the key takes precedence.

Reply:
I kept it for now